### PR TITLE
add validation to workload priority

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
When priorityClassName is set, priority should not be nil.
#### Which issue(s) this PR fixes:

Refer to https://github.com/kubernetes-sigs/kueue/pull/304#issuecomment-1204921639

### CODE REVIEW

File: apis/kueue/v1alpha1/workload_webhook.go [hunk 1]
```diff
@@ -95,37 +95,41 @@ func ValidateWorkload(obj *Workload) field.ErrorList {
...
 		msgs := validation.IsDNS1123Subdomain(obj.Spec.PriorityClassName)
 		if len(msgs) > 0 {
-			classNameField := specField.Child("priorityClassName")
 			for _, msg := range msgs {
-				allErrs = append(allErrs, field.Invalid(classNameField, obj.Spec.PriorityClassName, msg))
+				allErrs = append(allErrs, field.Invalid(specPath.Child("priorityClassName"), obj.Spec.PriorityClassName, msg))
 			}
 		}
+		if obj.Spec.Priority == nil {
+			allErrs = append(allErrs, field.Invalid(specPath.Child("priority"), obj.Spec.Priority, "priority should not be nil when priorityClassName once set"))
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/workload_webhook.go:103]:
```suggestion
			allErrs = append(allErrs, field.Invalid(specPath.Child("priority"), obj.Spec.Priority, "priority should not be nil when priorityClassName is set"))
```

File: apis/kueue/v1alpha1/workload_webhook.go [hunk 2]
```diff
@@ -95,37 +95,41 @@ func ValidateWorkload(obj *Workload) field.ErrorList {
...
+	allErrs = append(allErrs, validateAdmissionUpdate(newObj.Spec.Admission, oldObj.Spec.Admission, specPath.Child("admission"))...)
 
 	return allErrs
 }
 
 /// validateQueueNameUpdate validates that queueName is set once
-func validateQueueNameUpdate(new, old string) field.ErrorList {
+func validateQueueNameUpdate(new, old string, path *field.Path) field.ErrorList {
 	if len(old) == 0 {
 		return field.ErrorList{}
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/workload_webhook.go:124]: return nil

File: apis/kueue/v1alpha1/workload_webhook.go [hunk 3]
```diff
@@ -95,37 +95,41 @@ func ValidateWorkload(obj *Workload) field.ErrorList {
...
 	}
-	return apivalidation.ValidateImmutableField(new, old, field.NewPath("spec", "queueName"))
+	return apivalidation.ValidateImmutableField(new, old, path)
 }
 
 // validateAdmissionUpdate validates that admission is set once
-func validateAdmissionUpdate(new, old *Admission) field.ErrorList {
+func validateAdmissionUpdate(new, old *Admission, path *field.Path) field.ErrorList {
 	if old == nil {
 		return field.ErrorList{}
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/workload_webhook.go:132]: return nil

File: apis/kueue/v1alpha1/workload_webhook_test.go [hunk 4]
```diff
@@ -71,11 +72,22 @@ func TestValidateWorkload(t *testing.T) {
...
-			workload: testingutil.MakeWorkload(testWorkloadName, testWorkloadNamespace).PriorityClass("invalid_class").Obj(),
+			workload: testingutil.MakeWorkload(testWorkloadName, testWorkloadNamespace).
+				PriorityClass("invalid_class").
+				Priority(pointer.Int32(0)).
+				Obj(),
 			wantErr: field.ErrorList{
 				field.Invalid(specField.Child("priorityClassName"), "invalid_class", ""),
 			},
 		},
+		"should have priority once priorityClassName is set": {
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/workload_webhook_test.go:83]: add a test for nil priority when priorityClassName is not set

**Review by alculquicondor — COMMENTED** [review]
/approve

File: apis/kueue/v1alpha1/workload_webhook_test.go (hunk 4 above)
Comment (kerthcet) [reply apis/kueue/v1alpha1/workload_webhook_test.go:87]: Done

**Comment by kerthcet** [comment]
Squashed.

File: apis/kueue/v1alpha1/workload_webhook.go [hunk 5]
```diff
@@ -75,17 +75,17 @@ func (r *Workload) ValidateDelete() error {
 }
 
 func ValidateWorkload(obj *Workload) field.ErrorList {
-	var allErrs field.ErrorList
-	specField := field.NewPath("spec")
-	podSetsField := specField.Child("podSets")
+	allErrs := field.ErrorList{}
```
Comment (alculquicondor) [inline apis/kueue/v1alpha1/workload_webhook.go:78]: why change this? it's better to start with a nil array so that there are no unnecessary allocations.

Comment (kerthcet) [reply apis/kueue/v1alpha1/workload_webhook.go]: I can't remember why I change this, but it seems like unreasonable, reverted.

**Comment by kerthcet** [comment]
Flaky test is fixed by the second commit.

**Pull request merged**